test/unit_tests/unittest_scanner_controller.cpp

```diff
@@ -65,182 +68,151 @@ class ScannerControllerTest : public ::testing::Test
-
-  start_future.wait();
+  scanner_controller_.control_udp_client_.sendStopReply();
 }
 
-TEST_F(ScannerControllerTest, testStopRequestEvent)
+template <typename TestData>
+void ScannerControllerTest::sendMonitoringFrame(const TestData& test_data)
 {
-  EXPECT_CALL(scanner_controller_.state_machine_, processStopRequestEvent()).Times(1);
-
-  scanner_controller_.stop();
+  scanner_controller_.data_udp_client_.sendMonitoringFrame(test_data);
 }
 
-TEST_F(ScannerControllerTest, testStopRequestEventWithFutureUsage)
+void ScannerControllerTest::simulateUdpError(const std::string& msg)
 {
-  EXPECT_CALL(scanner_controller_.state_machine_, processStopRequestEvent()).Times(1);
-
-  std::future<void> stop_future = scanner_controller_.stop();
-
-  scanner_controller_.state_machine_.stopped_cb_();  // TODO needs to be real async?
-
-  stop_future.wait();
+  scanner_controller_.control_udp_client_.simulateError(msg);
 }
 
-TEST_F(ScannerControllerTest, testStartRequestSending)
+void ScannerControllerTest::simulateUdpTimeout(const std::string& msg)
 {
-  using ::testing::_;
-  using ::testing::Expectation;
-
-  StartRequest start_request(scanner_config_, 0);
-
-  Expectation control_udp_client_start_receiving =
-      EXPECT_CALL(scanner_controller_.control_udp_client_, startAsyncReceiving(_, _, _));
-  Expectation data_udp_client_start_receiving =
-      EXPECT_CALL(scanner_controller_.data_udp_client_, startAsyncReceiving());
-  EXPECT_CALL(scanner_controller_.control_udp_client_, write(start_request.toRawData()))
-      .After(control_udp_client_start_receiving, data_udp_client_start_receiving);
+  scanner_controller_.control_udp_client_.simulateTimeout(msg);
+}
 
-  scanner_controller_.sendStartRequest();
+template <typename TestData>
+LaserScan ScannerControllerTest::testDataToLaserScan(const TestData& test_data)
+{
+  const MaxSizeRawData raw_data = convertToMaxSizeRawData(test_data.hex_dump);
+  const auto num_bytes = 2 * test_data.hex_dump.size();
+  const MonitoringFrameMsg frame{ MonitoringFrameMsg::fromRawData(raw_data, num_bytes) };
+  return toLaserScan(frame);
 }
 
-TEST_F(ScannerControllerTest, testStopRequestSending)
+TEST_F(ScannerControllerTest, testStart)
 {
   using ::testing::_;
-  using ::testing::Expectation;
+  using ::testing::InSequence;
 
-  StopRequest stop_request;
+  const StartRequest start_request(scanner_config_, 0);
 
-  EXPECT_CALL(scanner_controller_.control_udp_client_, startAsyncReceiving(_, _, _)).Times(1);
-  EXPECT_CALL(scanner_controller_.control_udp_client_, write(stop_request.toRawData())).Times(1);
+  {
+    InSequence seq;
+    EXPECT_CALL(scanner_controller_.control_udp_client_, startAsyncReceiving(_, _, _)).Times(1);
+    EXPECT_CALL(scanner_controller_.data_udp_client_, startAsyncReceiving()).Times(1);
+    EXPECT_CALL(scanner_controller_.control_udp_client_, write(start_request.toRawData())).Times(1);
+  }
 
-  scanner_controller_.sendStopRequest();
+  auto start_future = scanner_controller_.start();
```

Comment:
```suggestion
  scanner_controller_.start();
```
is not possible?

Reply:
it is but we though testing for the promise to get fullfilled would be nicer

Reply:
Yes indeed. Now with `EXPECT_EQ(start_future.wait_for(std::chrono::seconds(0)), std::future_status::ready);` even better!